File: sdk/python/pepperx/websocket.py
"""Typed WebSocket client for PepperX.

Operations are correlated to responses by identifier, so many may be in flight on one connection.
"""

from __future__ import annotations

import asyncio
import base64
import json
import uuid
from typing import Any, Dict, List, Optional

import websockets

from .errors import PepperXError
from .models import (
    ApiError,
    Container,
    EnumerationQuery,
    EnumerationResult,
    ObjectMetadata,
    ObjectWriteResult,
    Statistics,
)

_DEFAULT_TIMEOUT = 100.0


class AsyncPepperXWebsocketClient:
    """Asynchronous client for the PepperX WebSocket surface.

    Call :meth:`connect` before issuing operations, then :meth:`close` when finished, or use the
    client as an async context manager.
    """

    def __init__(self, url: str = "ws://localhost:8002/", timeout: float = _DEFAULT_TIMEOUT) -> None:
        if not url:
            raise ValueError("url is required.")
        self.url = url
        self.timeout = timeout
        self._socket: Optional[Any] = None
        self._pending: Dict[str, asyncio.Future] = {}
        self._reader: Optional[asyncio.Task] = None

    async def __aenter__(self) -> "AsyncPepperXWebsocketClient":
        await self.connect()
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()

    @property
    def is_connected(self) -> bool:
        """Whether the client currently holds an open connection."""
        return self._socket is not None

    async def connect(self) -> None:
        """Connect to the server and start the receive loop."""
        if self._socket is not None:
            return
        self._socket = await websockets.connect(self.url, max_size=None)
        self._reader = asyncio.create_task(self._receive_loop())

    async def close(self) -> None:
        """Close the connection and fail any in-flight operations."""
        socket, self._socket = self._socket, None

        if self._reader is not None:
            self._reader.cancel()
            try:
                await self._reader
            except (asyncio.CancelledError, Exception):
                pass
            self._reader = None

        if socket is not None:
            await socket.close()

        for future in self._pending.values():
            if not future.done():
                future.set_exception(
                    PepperXError(ApiError.INTERNAL_ERROR, 499, "The client closed before a response arrived.")
                )
        self._pending.clear()

    # ------------------------------------------------------------- operations

    async def health(self) -> bool:
        """Return True when the node responds successfully."""
        response = await self._send("Health")
        return bool(response.get("Success"))

    async def create_container(self, name: str, tags: Optional[Dict[str, str]] = None) -> Container:
        """Create a container."""
        body: Dict[str, Any] = {"Name": name}
        if tags:
            body["Tags"] = tags
        response = await self._send("ContainerCreate", body=body)
        return Container.from_wire(self._require(response))

    async def read_container(self, name: str) -> Optional[Container]:
        """Read a container, or return None when it does not exist."""
        response = await self._send("ContainerRead", container=name)
        if response.get("StatusCode") == 404:
            return None
        return Container.from_wire(self._require(response))

    async def delete_container(self, name: str) -> None:
        """Delete an empty container."""
        self._raise(await self._send("ContainerDelete", container=name))

    async def write_object(
        self,
        container: str,
        key: str,
        data: bytes,
        content_type: Optional[str] = None,
        labels: Optional[List[str]] = None,
        tags: Optional[Dict[str, str]] = None,
        metadata_object: Optional[Any] = None,
        no_overwrite: bool = False,
    ) -> ObjectWriteResult:
        """Write an object."""
        body: Dict[str, Any] = {
            "DataBase64": base64.b64encode(data).decode("ascii"),
            "NoOverwrite": no_overwrite,
        }
        if content_type is not None:
            body["ContentType"] = content_type
        if labels:
            body["Labels"] = labels
        if tags:
            body["Tags"] = tags
        if metadata_object is not None:
            body["Object"] = metadata_object

        response = await self._send("ObjectWrite", container=container, key=key, body=body)
        return ObjectWriteResult.from_wire(self._require(response))

    async def read_object(self, container: str, key: str) -> Optional[bytes]:
        """Read an object's payload, or return None when it does not exist."""
        response = await self._send("ObjectRead", container=container, key=key)
        if response.get("StatusCode") == 404:
            return None
        self._raise(response)
        payload = response.get("DataBase64")
        return base64.b64decode(payload) if payload else b""

    async def read_object_metadata(self, container: str, key: str) -> Optional[ObjectMetadata]:
        """Read an object's metadata, or return None when it does not exist."""
        response = await self._send("ObjectReadMetadata", container=container, key=key)
        if response.get("StatusCode") == 404:
            return None
        return ObjectMetadata.from_wire(self._require(response))

    async def delete_object(self, container: str, key: str) -> bool:
        """Delete an object. Returns False when it did not exist."""
        response = await self._send("ObjectDelete", container=container, key=key)
        if response.get("StatusCode") == 404:
            return False
        self._raise(response)
        return True

    async def enumerate_objects(
        self, container: str, query: Optional[EnumerationQuery] = None
    ) -> EnumerationResult[ObjectMetadata]:
        """Enumerate or search objects within a container."""
        response = await self._send(
            "ObjectEnumerate", container=container, body=(query or EnumerationQuery()).to_wire()
        )
        payload = self._require(response)
        return EnumerationResult(
            success=payload.get("Success", True),
            max_results=payload.get("MaxResults", 100),
            continuation_token=payload.get("ContinuationToken"),
            end_of_results=payload.get("EndOfResults", True),
            total_records=payload.get("TotalRecords", 0),
            records_remaining=payload.get("RecordsRemaining", 0),
            objects=[ObjectMetadata.from_wire(o) for o in (payload.get("Objects") or [])],
        )

    async def statistics(self) -> Statistics:
        """Read aggregate statistics."""
        response = await self._send("AdminStats")
        return Statistics.from_wire(self._require(response))

    # ---------------------------------------------------------------- internal

    async def _send(
        self,
        operation: str,
        container: Optional[str] = None,
        key: Optional[str] = None,
        body: Optional[Any] = None,
    ) -> Dict[str, Any]:
        """Send one envelope and await its correlated response."""
        if self._socket is None:
            raise RuntimeError("The client is not connected; call connect() first.")

        request_id = uuid.uuid4().hex
        envelope: Dict[str, Any] = {"RequestId": request_id, "Operation": operation}
        if container is not None:
            envelope["Container"] = container
        if key is not None:
            envelope["Key"] = key
        if body is not None:
            envelope["Body"] = body

        loop = asyncio.get_running_loop()
        future: asyncio.Future = loop.create_future()
        self._pending[request_id] = future

        try:
            await self._socket.send(json.dumps(envelope))
            return await asyncio.wait_for(future, timeout=self.timeout)
        except asyncio.TimeoutError as exc:
            raise PepperXError(
                ApiError.INTERNAL_ERROR, 504, "Timed out waiting for a WebSocket response."
            ) from exc
        finally:
            self._pending.pop(request_id, None)

    async def _receive_loop(self) -> None:
        """Read responses and complete their pending futures."""
        try:
            while self._socket is not None:
                message = await self._socket.recv()
                envelope = json.loads(message)
                request_id = envelope.get("RequestId")
                if request_id and request_id in self._pending:
                    future = self._pending[request_id]
                    if not future.done():
                        future.set_result(envelope)
        except asyncio.CancelledError:
            raise
        except Exception as exc:
            for future in self._pending.values():
                if not future.done():
                    future.set_exception(
                        PepperXError(ApiError.INTERNAL_ERROR, 500, f"The WebSocket connection failed: {exc}")
                    )
            self._pending.clear()

    @staticmethod
    def _raise(response: Dict[str, Any]) -> None:
        """Raise when the server reported a failure."""
        if response.get("Success"):
            return

        error = response.get("Error") or {}
        raw = error.get("Error")
        try:
            error_type = ApiError(raw) if raw else ApiError.INTERNAL_ERROR
        except ValueError:
            error_type = ApiError.INTERNAL_ERROR

        status = response.get("StatusCode", 500)
        message = error.get("Message") or f"The operation failed with status {status}."
        raise PepperXError(error_type, status, message)

    @classmethod
    def _require(cls, response: Dict[str, Any]) -> Dict[str, Any]:
        """Raise on failure, then return the result payload."""
        cls._raise(response)
        result = response.get("Result")
        if result is None:
            raise PepperXError(
                ApiError.INTERNAL_ERROR, response.get("StatusCode", 500), "The server returned no result payload."
            )
        return result

File: sdk/js/src/websocket.ts
/**
 * Typed WebSocket client for PepperX.
 *
 * Requests are correlated to responses by identifier, so many operations may be in flight on one
 * connection. Works with the browser `WebSocket` global, or with a Node implementation such as `ws`
 * passed via options.
 */

import { PepperXError } from './errors.js';
import type {
  ApiError,
  Container,
  EnumerationQuery,
  EnumerationResult,
  ObjectMetadata,
  ObjectWriteResult,
  Statistics,
  WriteObjectOptions,
} from './types.js';

/** Minimal WebSocket surface the client depends on. */
export interface WebSocketLike {
  send(data: string): void;
  close(): void;
  addEventListener(type: string, listener: (event: any) => void): void;
}

/** Options accepted by the WebSocket client constructor. */
export interface PepperXWebsocketClientOptions {
  /** How long an operation waits for its correlated response, in milliseconds. Default 100000. */
  timeoutMs?: number;
  /** WebSocket constructor, for environments without a global one (for example Node with `ws`). */
  WebSocketImpl?: new (url: string) => WebSocketLike;
}

interface ResponseEnvelope {
  RequestId?: string;
  Success?: boolean;
  StatusCode?: number;
  Error?: { Error?: string; Message?: string } | null;
  Result?: Record<string, any> | null;
  DataBase64?: string | null;
}

function decodeBase64(value: string): Uint8Array {
  if (typeof atob === 'function') {
    const binary = atob(value);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
    return bytes;
  }
  return new Uint8Array(Buffer.from(value, 'base64'));
}

function encodeBase64(bytes: Uint8Array): string {
  if (typeof btoa === 'function') {
    let binary = '';
    for (const byte of bytes) binary += String.fromCharCode(byte);
    return btoa(binary);
  }
  return Buffer.from(bytes).toString('base64');
}

function toWireQuery(query: EnumerationQuery = {}): Record<string, unknown> {
  const body: Record<string, unknown> = {
    MaxResults: query.maxResults ?? 100,
    Skip: query.skip ?? 0,
    Ordering: query.ordering ?? 'CreatedDescending',
    CaseInsensitive: query.caseInsensitive ?? false,
  };
  if (query.prefix != null) body.Prefix = query.prefix;
  if (query.labels?.length) body.Labels = query.labels;
  if (query.tags && Object.keys(query.tags).length > 0) body.Tags = query.tags;
  if (query.containers?.length) body.Containers = query.containers;
  return body;
}

/**
 * Typed client for the PepperX WebSocket surface. Call {@link connect} before issuing operations.
 */
export class PepperXWebsocketClient {
  /** WebSocket URL of the target node. */
  readonly url: string;

  private readonly timeoutMs: number;
  private readonly WebSocketImpl: new (url: string) => WebSocketLike;
  private readonly pending = new Map<
    string,
    { resolve: (value: ResponseEnvelope) => void; reject: (reason: unknown) => void; timer: ReturnType<typeof setTimeout> }
  >();

  private socket: WebSocketLike | null = null;

  /**
   * @param url WebSocket URL, for example `ws://localhost:8002/`.
   * @param options Optional timeout and WebSocket implementation.
   */
  constructor(url = 'ws://localhost:8002/', options: PepperXWebsocketClientOptions = {}) {
    if (!url) throw new Error('url is required.');
    this.url = url;
    this.timeoutMs = options.timeoutMs ?? 100000;

    const impl = options.WebSocketImpl ?? (globalThis as any).WebSocket;
    if (!impl) {
      throw new Error('No WebSocket implementation is available; pass one via options.WebSocketImpl.');
    }
    this.WebSocketImpl = impl;
  }

  /** Whether the client currently holds an open connection. */
  get isConnected(): boolean {
    return this.socket !== null;
  }

  /** Connect to the server and start receiving responses. */
  async connect(): Promise<void> {
    if (this.socket) return;

    await new Promise<void>((resolve, reject) => {
      const socket = new this.WebSocketImpl(this.url);

      socket.addEventListener('open', () => {
        this.socket = socket;
        resolve();
      });

      socket.addEventListener('error', (event: any) => {
        reject(new Error(`WebSocket connection failed: ${event?.message ?? 'unknown error'}`));
      });

      socket.addEventListener('message', (event: any) => {
        this.onMessage(typeof event.data === 'string' ? event.data : String(event.data));
      });

      socket.addEventListener('close', () => {
        this.socket = null;
        this.failAll(new PepperXError('InternalError', 499, 'The WebSocket connection closed.'));
      });
    });
  }

  /** Close the connection and fail any in-flight operations. */
  async close(): Promise<void> {
    const socket = this.socket;
    this.socket = null;
    if (socket) socket.close();
    this.failAll(new PepperXError('InternalError', 499, 'The client closed before a response arrived.'));
  }

  /** Resolve true when the node responds successfully. */
  async health(): Promise<boolean> {
    const response = await this.send('Health');
    return response.Success === true;
  }

  /** Create a container. */
  async createContainer(name: string, tags?: Record<string, string>): Promise<Container> {
    const body: Record<string, unknown> = { Name: name };
    if (tags) body.Tags = tags;
    const data = this.require(await this.send('ContainerCreate', { body }));
    return {
      id: data.Id ?? '',
      name: data.Name ?? '',
      tags: data.Tags ?? {},
      objectCount: data.ObjectCount ?? 0,
      totalBytes: data.TotalBytes ?? 0,
      createdUtc: new Date(data.CreatedUtc ?? 0),
      lastUpdateUtc: new Date(data.LastUpdateUtc ?? 0),
    };
  }

  /** Delete an empty container. */
  async deleteContainer(name: string): Promise<void> {
    this.throwIfError(await this.send('ContainerDelete', { container: name }));
  }

  /** Write an object. */
  async writeObject(
    container: string,
    key: string,
    payload: Uint8Array,
    options: WriteObjectOptions = {},
  ): Promise<ObjectWriteResult> {
    const body: Record<string, unknown> = {
      DataBase64: encodeBase64(payload),
      NoOverwrite: options.noOverwrite ?? false,
    };
    if (options.contentType) body.ContentType = options.contentType;
    if (options.labels?.length) body.Labels = options.labels;
    if (options.tags && Object.keys(options.tags).length > 0) body.Tags = options.tags;
    if (options.metadataObject !== undefined) body.Object = options.metadataObject;

    const data = this.require(await this.send('ObjectWrite', { container, key, body }));
    return {
      extentId: data.ExtentId ?? '',
      key: data.Key ?? '',
      containerId: data.ContainerId ?? '',
      sizeBytes: data.SizeBytes ?? 0,
      sha256: data.Sha256 ?? '',
      contentType: data.ContentType ?? null,
      replaced: data.Replaced ?? false,
    };
  }

  /** Read an object's payload, or resolve null when it does not exist. */
  async readObject(container: string, key: string): Promise<Uint8Array | null> {
    const response = await this.send('ObjectRead', { container, key });
    if (response.StatusCode === 404) return null;
    this.throwIfError(response);
    return response.DataBase64 ? decodeBase64(response.DataBase64) : new Uint8Array();
  }

  /** Read an object's metadata, or resolve null when it does not exist. */
  async readObjectMetadata(container: string, key: string): Promise<ObjectMetadata | null> {
    const response = await this.send('ObjectReadMetadata', { container, key });
    if (response.StatusCode === 404) return null;
    const data = this.require(response);
    return {
      key: data.Key ?? '',
      extentId: data.ExtentId ?? '',
      containerId: data.ContainerId ?? '',
      containerName: data.ContainerName ?? null,
      sizeBytes: data.SizeBytes ?? 0,
      sha256: data.Sha256 ?? '',
      contentType: data.ContentType ?? null,
      labels: data.Labels ?? [],
      tags: data.Tags ?? {},
      metadataObject: data.Object ?? null,
      hasMetadataObject: data.HasMetadataObject ?? false,
      createdUtc: new Date(data.CreatedUtc ?? 0),
    };
  }

  /** Delete an object. Resolves false when it did not exist. */
  async deleteObject(container: string, key: string): Promise<boolean> {
    const response = await this.send('ObjectDelete', { container, key });
    if (response.StatusCode === 404) return false;
    this.throwIfError(response);
    return true;
  }

  /** Enumerate or search objects within a container. */
  async enumerateObjects(
    container: string,
    query: EnumerationQuery = {},
  ): Promise<EnumerationResult<ObjectMetadata>> {
    const data = this.require(
      await this.send('ObjectEnumerate', { container, body: toWireQuery(query) }),
    );
    return {
      success: data.Success ?? true,
      maxResults: data.MaxResults ?? 100,
      continuationToken: data.ContinuationToken ?? null,
      endOfResults: data.EndOfResults ?? true,
      totalRecords: data.TotalRecords ?? 0,
      recordsRemaining: data.RecordsRemaining ?? 0,
      objects: (data.Objects ?? []).map((o: Record<string, any>) => ({
        key: o.Key ?? '',
        extentId: o.ExtentId ?? '',
        containerId: o.ContainerId ?? '',
        containerName: o.ContainerName ?? null,
        sizeBytes: o.SizeBytes ?? 0,
        sha256: o.Sha256 ?? '',
        contentType: o.ContentType ?? null,
        labels: o.Labels ?? [],
        tags: o.Tags ?? {},
        metadataObject: o.Object ?? null,
        hasMetadataObject: o.HasMetadataObject ?? false,
        createdUtc: new Date(o.CreatedUtc ?? 0),
      })),
    };
  }

  /** Read aggregate statistics. */
  async statistics(): Promise<Statistics> {
    const data = this.require(await this.send('AdminStats'));
    return {
      containerCount: data.ContainerCount ?? 0,
      objectCount: data.ObjectCount ?? 0,
      totalBytes: data.TotalBytes ?? 0,
      containers: [],
      storageTotalBytes: data.StorageTotalBytes ?? 0,
      storageFreeBytes: data.StorageFreeBytes ?? 0,
      databaseSizeBytes: data.DatabaseSizeBytes ?? 0,
      nodes: [],
    };
  }

  // -------------------------------------------------------------------- private

  private send(
    operation: string,
    options: { container?: string; key?: string; body?: unknown } = {},
  ): Promise<ResponseEnvelope> {
    if (!this.socket) throw new Error('The client is not connected; call connect() first.');

    const requestId = `${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 10)}`;
    const envelope: Record<string, unknown> = { RequestId: requestId, Operation: operation };
    if (options.container !== undefined) envelope.Container = options.container;
    if (options.key !== undefined) envelope.Key = options.key;
    if (options.body !== undefined) envelope.Body = options.body;

    return new Promise<ResponseEnvelope>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(requestId);
        reject(new PepperXError('InternalError', 504, 'Timed out waiting for a WebSocket response.'));
      }, this.timeoutMs);

      this.pending.set(requestId, { resolve, reject, timer });
      this.socket!.send(JSON.stringify(envelope));
    });
  }

  private onMessage(raw: string): void {
    let envelope: ResponseEnvelope;
    try {
      envelope = JSON.parse(raw);
    } catch {
      return;
    }

    const id = envelope.RequestId;
    if (!id) return;

    const entry = this.pending.get(id);
    if (!entry) return;

    clearTimeout(entry.timer);
    this.pending.delete(id);
    entry.resolve(envelope);
  }

  private failAll(error: unknown): void {
    for (const entry of this.pending.values()) {
      clearTimeout(entry.timer);
      entry.reject(error);
    }
    this.pending.clear();
  }

  private throwIfError(response: ResponseEnvelope): void {
    if (response.Success) return;
    const errorType = (response.Error?.Error ?? 'InternalError') as ApiError;
    const status = response.StatusCode ?? 500;
    const message = response.Error?.Message ?? `The operation failed with status ${status}.`;
    throw new PepperXError(errorType, status, message);
  }

  private require(response: ResponseEnvelope): Record<string, any> {
    this.throwIfError(response);
    if (!response.Result) {
      throw new PepperXError('InternalError', response.StatusCode ?? 500, 'The server returned no result payload.');
    }
    return response.Result;
  }
}
